### PR TITLE
PathCORE methods documentation and transfer from bitbucket

### PR DESCRIPTION
**Pull request by kathyxchen**
This pull request looks very large--I want to start off by making it clear that the majority of the methods in this PR have been reviewed previously by @tj8901nm, but in contexts that were not PathCORE-specific. 

Please let me know if you think someone else would be more appropriate as a reviewer on this PR. Thanks!!

**NEEDS REVIEW:**
- Package-specific files: 
  - `README.rst` will have to be updated in a small follow-up PR once the PathCORE analysis repository (specific to the PathCORE paper case studies) to link to usage examples. Outside of the lack of examples, what else am I missing here?
  - `setup.py` same question as the previous point.
  - `__init__.py` same question
- `network_permutation_test.py`: I had some parts of this on bitbucket as well, but the components were more spread out (harder to directly link to something like I did below). This is the main thing that needs to be reviewed here!

**PREVIOUSLY REVIEWED:**
- `feature_overrepresentation_analysis.py`: original file [here](https://bitbucket.org/greenelab/eadage/src/d8a98586e0f448a5b7d497f5516cc4e241635361/ensemble_construction/pathway_coverage_no_crosstalk/pathway_coverage_no_crosstalk.py?at=default&fileviewer=file-view-default). This is exactly the same, minus documentation updates, to the file in the eADAGE repository. The major change is that the [crosstalk correction](https://github.com/kathyxchen/crosstalk-correction) procedure has been removed from the module and exists as its own Python package.

- `network.py`: original file [here](https://bitbucket.org/greenelab/pathway-network-construction/src/86d49074ae238f56d0ba3fe3d6f463c5bb068cda/network.py?at=default&fileviewer=file-view-default). Previously, we had considered creating data structures that handled 2 or even 3 types of edges, but this has been simplified to a single edge type. The `CoNetwork` class that exists in this module is a simplification of the original file. The implementation has not changed, but I've updated the documentation and function names for clarity.

### CODE REVIEW

File: pathcore/__init__.py [hunk 1]
```diff
@@ -0,0 +1,14 @@
+from .feature_pathway_overrepresentation import \
+    pathway_enrichment_no_overlap_correction, \
+    pathway_enrichment_with_overlap_correction, \
+    single_side_pathway_enrichment
+from .network import CoNetwork
+from .network_permutation_test import aggregate_permuted_network, \
+        network_edges_permutation_test
+
+__all__ = ["CoNetwork",
```
Comment (dhimmel) [inline pathcore/__init__.py]: My understanding is that the only purpose of `__all__` is to enable `from pathcore import *` ([doc](https://docs.python.org/3/tutorial/modules.html#importing-from-a-package)). Unless this API functionality is required, I'd discourage it. Import all is a [bad practice](https://stackoverflow.com/a/2386740/4651668) and something I'd choose not to support.

File: setup.py [hunk 2]
```diff
@@ -0,0 +1,32 @@
...
+
+# Allow setup.py to be run from any path
+os.chdir(os.path.normpath(os.path.join(os.path.abspath(__file__), os.pardir)))
+
+setup(
+    name="PathCORE",
+    version="1.0.0",
+    packages=["pathcore"],
+    include_package_data=True,
+    license="LICENSE",
```
Comment (dhimmel) [inline setup.py]: `BSD-3-Clause`

File: setup.py [hunk 3]
```diff
@@ -0,0 +1,32 @@
...
+    author_email="team@greenelab.com",
+    classifiers=[
+        "Intended Audience :: Developers",
+        "License :: OSI Approved :: BSD License",
+        "Operating System :: OS Independent",
+        "Programming Language :: Python",
+        "Programming Language :: Python :: 3",
+        "Topic :: Scientific/Engineering :: Bio-Informatics",
+    ],
+    install_requires=["crosstalk-correction", "numpy", "pandas",
```
Comment (dhimmel) [inline setup.py]: best to put each requirement on its own line for better git diffing.

File: setup.py [hunk 4]
```diff
@@ -0,0 +1,32 @@
+import os
+from setuptools import setup
+
+
+with open(os.path.join(os.path.dirname(__file__), "README.rst")) as readme:
```
Comment (dhimmel) [inline setup.py]: getting the directory that `setup.py` is in is repeated below. Maybe do once and use in both places?

File: pathcore/network_permutation_test.py [hunk 5]
```diff
@@ -0,0 +1,185 @@
...
+        permuted_network = pathcore_network.permute_pathways_across_features()
+        if not aggregate_permuted_network:
+            aggregate_permuted_network = permuted_network
+        else:
+            aggregate_permuted_network.aggregate(permuted_network)
+    return aggregate_permuted_network
+
+
+def network_edges_permutation_test(observed_network,
+                                   permuted_networks,
```
Comment (dhimmel) [inline pathcore/network_permutation_test.py:44]: Where are these permuted networks generated?

Comment (kathyxchen) [reply pathcore/network_permutation_test.py:44]: Hmm.. so in the PathCORE-analysis repository [that will hopefully exist soon], I have a script that shows how `network_edges_permutation_test` is used. 
In the next commit on this PR
- I'll point to that in the README
- I'll also make it clear in this function's documentation that `permuted_networks` is generated by calling the [`CoNetwork`'s `permute` function](https://github.com/greenelab/PathCORE/pull/1/files/35adf7eb0700681ceff49d2ba819191b77807f8e#diff-1ed75b07a3d5795a002b33a5fb4c1337R189) _N_ times and collecting the output in a list.

Would that make things more clear?

Comment (dhimmel) [reply pathcore/network_permutation_test.py:44]: Ah got it. I missed the `permute_pathways_across_features` function.

File: pathcore/network_permutation_test.py [hunk 6]
```diff
@@ -0,0 +1,185 @@
...
+    """This method handles the case where multiple observed networks are
+    generated (e.g. from models produced by different random seed
+    initializations of an unsupervised feature construction algorithm).
+    We handle analysis of multiple networks by aggregating them; likewise,
+    we require that the permutation test generates N aggregate permuted
+    networks.
+
+    Parameters
+    -----------
+    observed_networks : list(CoNetwork)
```
Comment (kurtwheeler) [inline pathcore/network_permutation_test.py:22]: Python 3 supports type hints. If you:
```
from typing import List
from netword import CoNetwork
```
then you can change the function definition to:
```
def aggregate_permuted_network(observed_networks: List[CoNetwork]) -> CoNetwork:
```
which is unfortunately just a bit too long of a line... And people say 80 characters is plenty.

File: pathcore/network_permutation_test.py [hunk 7]
```diff
@@ -0,0 +1,185 @@
...
+        observed_weight = edge_obj.weight
+        weight_distribution = edge_weight_distributions[edge_id]
+        eq_or_above_observed = sum(weight_distribution[observed_weight:])
+        pvalue = eq_or_above_observed / float(n_permutations)
+        pvalues.append(pvalue)
+        edge_expected_weight.append((edge_id, _edge_expected_weight(
+            weight_distribution[1:])))
+
+    below_alpha, qvalues, _, _ = multipletests(
+        pvalues, alpha=alpha, method="fdr_bh")
```
Comment (kurtwheeler) [inline pathcore/network_permutation_test.py]: What is `fdr_bh`?

**Review by kurtwheeler — APPROVED** [review]
I agree with the comments @dhimmel made. I left only a couple myself and it's certainly optional whether you choose to do anything with them or not.

File: pathcore/network_permutation_test.py (hunk 6 above)
Comment (kathyxchen) [reply pathcore/network_permutation_test.py:22]: Oh I didn't know that! Had only converted scripts to Python 3 recently - I will look into it! Might add "include type hints" as an issue on this & other relevant repos once everything has been checked in :+1: Thanks!

File: pathcore/network_permutation_test.py (hunk 7 above)
Comment (kathyxchen) [reply pathcore/network_permutation_test.py]: false discovery rate, benjamini & hochberg. will add as a comment above this line

**Review by dhimmel — APPROVED** [review]
I'm not planning on looking at this PR again unless specifically requested to look over something else.

Nice work :black_flag:

**Pull request merged**